pkg/storage/cacher.go

```diff
@@ -497,14 +497,34 @@ func (c *cacheWatcher) stop() {
 	}
 }
 
+var timerPool sync.Pool
+
 func (c *cacheWatcher) add(event watchCacheEvent) {
-	t := time.NewTimer(5 * time.Second)
+	// Try to send the event immediately, without blocking.
+	select {
+	case c.input <- event:
+		return
+	default:
+	}
+
+	// OK, block sending, but only for up to 5 seconds.
+	// cacheWatcher.add is called very often, so arrange
+	// to reuse timers instead of constantly allocating.
+	const timeout = 5 * time.Second
+	t, ok := timerPool.Get().(*time.Timer)
+	if ok {
+		t.Reset(timeout)
```

Comment:
Fixed. Now every timer in the pool is both stopped and known not to have a channel event waiting.
In my earlier version there was only a possible problem if the event send took nearly exactly 5 seconds, so that the select could pick the c.input <- event case but then have the timer go off before the immediately following t.Stop. In that case there would be a leftover timer event in t.C that would be found when the timer was reused. In practice this was very unlikely to happen, and it's very difficult to force in a test. In some future version of Go I hope we'll be able to put the old time.After code back and have it do all the right things as far as efficiency instead of us doing it. But this will help for now.